Remotion/ObjectBinding/Web/UI/Controls/Validation/BocListValidationFailureRepository.cs

```diff
@@ -0,0 +1,265 @@
+﻿// This file is part of the re-motion Core Framework (www.re-motion.org)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+//
+// The re-motion Core Framework is free software; you can redistribute it
+// and/or modify it under the terms of the GNU Lesser General Public License
+// as published by the Free Software Foundation; either version 2.1 of the
+// License, or (at your option) any later version.
+//
+// re-motion is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+//
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Remotion.Collections;
+using Remotion.ObjectBinding.Validation;
+using Remotion.Utilities;
+
+namespace Remotion.ObjectBinding.Web.UI.Controls.Validation
+{
+  /// <summary>
+  /// Repository which contains all <see cref="BusinessObjectValidationFailure"/> belonging to a BocList.
+  /// </summary>
+  public class BocListValidationFailureRepository: IBocListValidationFailureRepository
+  {
+    private class HandleableFailure
+    {
+      public BusinessObjectValidationFailure Failure { get; }
+
+      public bool Handled { get; private set; }
+
+      public HandleableFailure (BusinessObjectValidationFailure failure)
+      {
+        Failure = failure;
+        Handled = false;
+      }
+
+      public void MarkAsHandled ()
+      {
+        Handled = true;
+      }
+    }
+
+    private class RowFailureContainer
+    {
+      private readonly List<HandleableFailure> _rowFailures = new List<HandleableFailure>();
+      private readonly Dictionary<IBocColumnDefinitionWithValidationSupport, List<HandleableFailure>> _cellFailures = new Dictionary<IBocColumnDefinitionWithValidationSupport, List<HandleableFailure>>();
+
+      public void AddRowFailures (IEnumerable<BusinessObjectValidationFailure> failures)
+      {
+        _rowFailures.AddRange(failures.Select(e => new HandleableFailure(e)));
+      }
+
+      public void AddCellFailures (IBocColumnDefinitionWithValidationSupport columnDefinition, IEnumerable<BusinessObjectValidationFailure> failures)
+      {
+        List<HandleableFailure> cellFailureForRow;
+        if (_cellFailures.TryGetValue(columnDefinition, out var existingFailureList))
+        {
+          cellFailureForRow = existingFailureList;
+        }
+        else
+        {
+          cellFailureForRow = new List<HandleableFailure>();
+          _cellFailures[columnDefinition] = cellFailureForRow;
+        }
```

Comment:
Possible simplification. May or may not feel better.
```suggestion
        List<HandleableFailure> cellFailureForRow;
        if (!_cellFailures.TryGetValue(columnDefinition, out var cellFailureForRow))
        {
          cellFailureForRow = new List<HandleableFailure>();
          _cellFailures.Add(columnDefinition, cellFailureForRow);
        }
```
Eitehr way, we should use Add not indexer for setting the value. Add has the semantic to fail if the item exists, indexer will overwrite. Since we want to add, not overwrite, the `Add` api should be used for expressiveness.